pydantic_settings/sources.py

```diff
@@ -1725,6 +1738,47 @@ def _read_file(self, file_path: Path) -> dict[str, Any]:
             return yaml.safe_load(yaml_file) or {}
 
 
+class AzureKeyVaultSettingsSource(EnvSettingsSource):
+    _url: str
+    _credential: TokenCredential  # type: ignore
+    _secret_client: SecretClient  # type: ignore
+
+    def __init__(
+        self,
+        settings_cls: type[BaseSettings],
+        url: str,
+        credential: TokenCredential,  # type: ignore
+        env_parse_none_str: str | None = None,
+        env_parse_enums: bool | None = None,
+    ) -> None:
+        import_azure_key_vault()
+        self._url = url
+        self._credential = credential
+        super().__init__(
+            settings_cls,
+            case_sensitive=True,
+            env_prefix='',
+            env_nested_delimiter='--',
+            env_ignore_empty=False,
+            env_parse_none_str=env_parse_none_str,
+            env_parse_enums=env_parse_enums,
+        )
+
+    def _load_env_vars(self) -> Mapping[str, str | None]:
```

Comment:
Couldn't we use a lazy mapping, something like

```py
from collections import Mapping


class AzureMapping(Mapping):
    def __init__(self, secret_client):
        self._secret_client = secret_client
        self._keys = self._secret_client.get_names()

    def __getitem__(self, __key):
        return self._secret_client.get_secret(__key)

    def __len__(self):
        return len(self._keys)

    def __iter__(self):
        return iter(self._keys)
```

Reply:
why it doesn't work. the whole point of lazy mapping is prevent looping over all keys in keyvalult and only fetch the one that we need.

Reply:
@hramezani When I execute the code and instance `ApplicationSettings`, `_load_env_vars` is called once per key, so all of them are fetched.

Reply:
It seems "lazy mapping" means only fetching the keys we need from Key Vault, not preventing the fetch action (lazy loading).

Ok, the new code works well, but in this example I get an error: `SettingsError: error getting value for field "sql_server" from source "AzureKeyVaultSettingsSource"`.  Specifically, the error happens because `__getitem__`
 is called with the key `SqlServer` instead of `SqlServer--Password.`

Reply:
I think it is working as expected. what is the of key `SqlServer` in key-vault? the below error: 

`SettingsError: error getting value for field "sql_server" from source "AzureKeyVaultSettingsSource"`

You defined `sql_server: SqlServer = Field(..., alias="SqlServer")` which is a nested field. so , you have to define a valid json string(because `pydantic-settings` considers this field as a complex field and tries to json parse it) for it in key-vault. the key has to be `SqlServer` and the value has to be e.g `'{"Password": "key"}'`

Reply:
In Key Vault, I have the key `SqlServer--Password`. In this Azure service, the nesting is always done with the `--` separator. Key Vault secrets store secrets/passwords (their keys can only have alphanumeric characters and dashes), can be referenced in several ways (App Configuration references, App Configuration SDK, Key Vault references, Key Vault SDK, ASP.NET Core integration, pydantic-settings, etc.), and are shared among applications, so they mustn't contain a JSON attached to a Pydantic model, they must contain the raw secret. And we should be able to map groups of secrets into a Pydantic model (SqlServer--User, SqlServer--Password, SqlServer--...) without modifying the password.

Using the branch code without `AzureKeyVaultMapping`, this Azure service works in its proper way.

Reply:
you probably have value for `SqlServer` in your key-vault. then `pydantic-settings` fetched that one instead of `SqlServer--Password`. that's why I suggested changing the value of `SqlServer` to a valid json string.

So, I think the problem is because you have both `SqlServer` and `SqlServer--Password` defined in key-vault. then pydantic-settings tries to get the value first from `SqlServer` and fails.

This is not related to the key vault settings source, if you have both `SqlServer` and `SqlServer--Password` env variable, you will get the same error.

This is not a problem. you can fix it by:

- Remove the `SqlServer` key or convert it to a valid json string
- Or, change the alias for `sql_server`

Reply:
Having only the secret `SqlServer--Password` in Key Vault, `__getitem__` is called twice: one with `SqlServer` and another one with `SqlServer--Password`.

Reply:
Having only `SqlServer--Password` in Key Vault

![image](https://github.com/user-attachments/assets/490ba54f-a898-4350-8868-5a1361012568)

it throws an exception because `__getitem__` is called twice: once with `SqlServer` and another one with `SqlServer--Password`. I've had to add a try...except.

I've pushed the changes if you want to try it. The tests are failing too.

Reply:
> it throws an exception because __getitem__ is called twice: once with SqlServer and another one with SqlServer--Password. I've had to add a try...except.

This is the correct behavior. I think you have to replace the `return None` in `except` block with a `raise KeyError` to make it similar to a real dict.

> I've pushed the changes if you want to try it. The tests are failing too.

The tests are failing because you mocked the `get_secret` to return `SecretValue` for all the keys. so, it returns `SecretValue` for `SqlServer` and this is not a valid json string. that's why you get a parsing error. you have to change the test in a way to only return `SecretValue` for secret keys that we want and `ResourceNotFoundError` for others

I will try it later

Reply:
Done!